### ISSUE TITLE
Rakudo incorrectly calls methods on Match subclasses (eg Grammars) instead of on Match itself

### ISSUE THREAD

**Issue by codesections** [rakudo/rakudo#5468]
When looking into #5443, I (incorrectly) believed that issue to be caused by a broader problem with how `Match` methods are being called internally; based on belief, I changed the title to reflect the bigger issue.  I was mistaken; #5443 was actually triggered by a bug in how action methods were called.  Thus, I changed #5443's title back and am opening this issue to track the broader problem.

Here's the problem:  Because regexes (including tokens/rules) are methods on a grammar, they share a namespace with regular methods.  (This design decision has pros and cons, but this issue isn't about design; it's about fixing bugs in the current implementation.  If we want to change the design, we'd need to discuss that separately + wait for a language release (it'd be a breaking change).)

The shared namespace means that Rakudo-internal code dealing with `Grammar`s/`Match`s needs to be very explicit about what methods it's calling to avoid incorrectly calling user-defined regexes.  For example, `Match.raku` currently calls `self.from` – that is, it calls the last `from` method in the `mro`, which could easily be some variety of regex.  Instead, `Match.raku` should call `self.Match::from`.

That's just one example; there are likely dozens of similar changes that will need to be made to ensure that Rakudo-internal code does not incorrectly call inherited regexes when it means to call a specific method.  Once these bugs are fixed/this issue is closed, we should be able to remove the [traps entry](https://docs.raku.org/language/traps#Using_certain_names_for_rules/token/regexes) on name collisions between regexes and method.

Related issues: Raku/nqp#367, rakudo/rakudo#1222
Related IRC discussion: [2023-11-08](http://irclogs.raku.org/raku/2023-11-08.html#21:05-0001), [probably others; feel free to mention them and I'll add links]

**Comment by raiph** [comment]
@codesections,

The traps entry link 404s. Should presumably be https://docs.raku.org/language/traps#Using_certain_names_for_rules/token/regexes

I think this reflects a misunderstanding. There was a nasty bug due to calling methods on Mu that Liz just fixed. The name clash there is something completely different from the intentional design that rules are methods and follow standard method resolution, er, rules. That's not a name clash, that's intentional. I've commented further on this in the PS 401 issue (which I read first).

**Comment by codesections** [comment]
> The traps entry link 404s.

Thanks, fixed.

> That's not a name clash, that's intentional. I've commented further on this in the PS 401 issue

I replied to your comments in raku/problem-solving#401.  But I agree that the shared MRO is intentional; I nevertheless think that there are some (many?) places in Rakudo's current code where we didn't think carefully enough about that shared MRO and thus didn't write code that consistently calls the method it intended to call in the presence of subclass methods (including regexes) with the same name.

**Comment by lizmat** [comment]
I'm considering looking into lifting the `NQPMatchRole` from NQP into Raku land, and then change the Grammar class to not even inherit from `Mu`, but only doing that role.  With the internal methods of `NQPMatchRole` renamed to be much less likely to cause collisions.

Does that sound like a plan?  Am I missing something that would make this impossible?

**Comment by codesections** [comment]
@lizmat 
I'm slightly confused by your proposal. Right now, grammars and their matches (which are just grammars of the same type) have the following mro's:

```raku
grammar G { token TOP { . } }
say G.^mro;   # OUTPUT: «((G) (Grammar) (Match) (Capture) (Cool) (Any) (Mu))»
my $m = G.parse('f');
say $m.^mro; # OUTPUT: «((G) (Grammar) (Match) (Capture) (Cool) (Any) (Mu))»
```

With the setup you're suggesting, would both of those have an empty mro?  If so, does that mean that grammar matches wouldn't have `Match` methods for displaying/gisting?  And if they're not `Cool`, then the wouldn't be coerciable into `Str`s either, right?

Or would the idea be to re-implement (many of) those methods for grammars (or move them to the new `MatchRole`?  If so, that might be a fair bit to re-implement.

Or would grammars not inherit from `Match` but the result of `G.parse` still would?

**Comment by lizmat** [comment]
Technically, the `G` grammar would only have the `Grammar` class as its parent.

The `$m` would be a `Match` object as before, nothing changes there.

> Or would grammars not inherit from Match but the result of G.parse still would?

Indeed.

**Comment by niner** [comment]
On Sonntag, 12. November 2023 12:09:53 CET Elizabeth Mattijsen wrote:
> Technically, the `G` grammar would only have the `Grammar` class as its
> parent.

That would mean that you cannot even put it into a plain lexical variable or 
pass it as an untyped argument. While Raku looks very dynamically typed in 
truth type constraints are everywhere. It's just that you don't notice a Mu or 
Any type constraint when everything descends from those.

**Comment by lizmat** [comment]
Actually, I think that will work out, because a `Mu` constraint basically means no constraint.  How else would we be able to pass NQP objects around?

**Comment by niner** [comment]
But parameter type constraints are Any by default:
```
rakudo -e 'sub foo($a) { }(NQPMatchRole)'
Died with X::TypeCheck::Binding::Parameter
```

**Comment by lizmat** [comment]
Indeed.   But that is easily fixed with: `sub foo(Mu $a) { }(NQPMatchRole)`

**Comment by codesections** [comment]
> The `$m` would be a `Match` object as before, nothing changes there.

I'm still confused.  Right now, if I have a grammar named `G`, each time one of `G`'s rules matches, it returns a match of type `G` (which inherits from `Grammar`, `Match`, `Capture`, `Cool`, `Any`, and `Mu`).  And then `G`'s methods are called again on that `G` to proceed with the match.  So even if `Grammar`s don't inherit from anything, don't we end up with exactly the same issues with the intermediate matches that are returned as `G.parse` is in progress (since they're both `G`s and `Match`es)?

**Comment by lizmat** [comment]
Hmmmm  actually, you got a good point there.  I had only considered the final return value of `.parse`.   Hmmmm   this will indeed kill the cunning plan I had.  *sigh*   :-)

**Comment by lizmat** [comment]
Looks like this is the reason why a custom `<ws>` token doesn't work:

https://github.com/Raku/nqp/blob/main/src/QRegex/NFA.nqp#L376-L383

**Comment by lizmat** [comment]
This can now be fixed (I believe) by installing the [actions](https://raku.land/zef:lizmat/actions) distribution, and create the action methods inside an `actions` package.

**Comment by lizmat** [comment]
Yeah, I think an "actions" package type would make sense in core.
